### ISSUE TITLE
Debugging SteamOS is a lot harder than it should be

### ISSUE THREAD

**Issue by unclejack** [ValveSoftware/SteamOS#948]
#### Your system information

* Steam client version: latest stable
* SteamOS version: 3.4.2
* Opted into Steam client beta?: No
* Opted into SteamOS beta?: No 
* Have you checked for updates in Settings > System?: Yes

#### Please describe your issue in as much detail as possible:
SteamOS gets stuck with the Steam Deck logo on screen after starting the Linux kernel. The fans spin. Nothing else happens.
This happens often when SteamOS is booted from an external drive.

Steam OS' default configuration makes it rather difficult to debug. There's no support for serial console or other such debugging mechanisms. Changes need to be made to the OS to at least be able to debug these issues.

Another unpleasant class of issues is the one related to updating the Steam client when the OS starts the gamescope session. If it fails to connect the first time or it thinks it's offline, it displays a dialog box which complains that the Steam Deck is offline. There's no way to click the button, not even with a mouse and a keyboard connected.

My Steam Deck has been stuck with its fans spinning for about 10 minutes with the Steam Deck logo on the screen.

These issues make using SteamOS more frustrating than a regular Linux distribution. It certainly makes it annoying when trying to debug the system and make improvements.

#### Steps for reproducing this issue:

1. Install SteamOS 3.4.2 on an external disk
2. Boot SteamOS from the external disk
3. It may or may not boot once, let you to log into your account, install some games
4. If it did boot and start the Steam client in the gamescope session, shutdown
5. boot again to get stuck at the Steam Deck logo OR just look at the Steam Deck logo in step 3 if it didn't boot the first time either

**Comment by bertogg** [comment]
I'm not sure if I understand the use case here: are you installing the OS on an external disk and then booting the deck from that disk rather than from the internal one?

**Comment by unclejack** [comment]
> I'm not sure if I understand the use case here: are you installing the OS on an external disk and then booting the deck from that disk rather than from the internal one?

Yes, that's what I was trying to do. I need to debug via serial to figure out what's going on. I was under the impression that only the gamescope session is affected by some issues. It turns out that there are more issues which prevent SteamOS from booting properly from an external disk.

The goal is to be able to run a stable vanilla OS on the internal SSD and a separate SteamOS install on an external drive. The SteamOS instance running from the external drive is meant to be used for debugging and for testing improvements. One such example are the commits mentioned in https://github.com/ValveSoftware/SteamOS/issues/954 to improve responsiveness with the current kernel tree.

**Comment by bertogg** [comment]
You know that the internal SSD has two copies of the OS, right? If you want to test a different kernel you can do it on the non-active root filesystem. Press the (●●●) button when powering up the machine to get to the boot loader and decide which one to boot.

**Comment by unclejack** [comment]
> You know that the internal SSD has two copies of the OS, right? If you want to test a different kernel you can do it on the non-active root filesystem. Press the (●●●) button when powering up the machine to get to the boot loader and decide which one to boot.

I know it has two copies of the OS. I'll use that button to get to the boot loader. The issue I was running into was a blank screen without any further progress. I'll try serial debugging to figure out what's going on.

Thanks for the reply and for the help.

**Comment by unclejack** [comment]
@bertogg: The Steam Deck is having difficulties booting from the external drive again. I'm not able to get it into the grub menu either. Perhaps there's another way to reach the grub prompt? That would at least make it possible to enable debugging without making permanent changes or to select a different kernel on boot.

**Comment by bertogg** [comment]
If you hold the Volume Down button while powering up the machine you'll get to the BIOS boot menu, but I'm not sure that that will solve your problem.

Why can't you see the grub prompt? What happens instead? Did you modify the OS or are you simply connecting an external drive?

**Comment by unclejack** [comment]
I was trying to get to the grub prompt when booting from the external drive. I know the volume down + power button combination which is used to reach the boot menu from the firmware.

The goal is to try to investigate why the system doesn't boot properly. If there's some issue with the OS, it'd be cool to figure it out and be able to debug kernels (5.13 with backports, 6.1).

**Comment by lostgoat** [comment]
Booting from an external drive isn't something we support, so any adventures in that direction are all "at your own risk".

However, if you are booting from USB to try to have an easy way to recover your system in case of a bad kernel, you might find SteamOS's A/B partitioning scheme useful. You can access the relevant boot menu to boot into the alternate OS partition by holding the '...' button and then pressing the power button.

That way if you install a bad kernel or end up in any bad situation by editing OS packages you can recover the system by just booting to your previous OS partition. Just make sure not to mess both of them at the same time.

**Comment by unclejack** [comment]
@lostgoat: That was helpful. I'm able to reach the bootloader. This has also helped me figure out that the kernel boots and systemd does in fact start the system when the boot process gets stuck when booting from an external drive. I'll try to figure out why that happens sometimes.

Regarding the support, that's fine. I'll debug on my own. This has been helpful.
